src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -1064,24 +1035,23 @@ private enum CacheStatus {
 			}
 
 			// get a filestream from the pool, or create one if the pool is empty.
-			if (_fileStreams.TryTake(out item))
-				return item;
+			if (_fileStreams.TryTake(_handle, &CreateFileStreamWorkItem) is { } fileStreamWorkItem)
+				return fileStreamWorkItem;
 
-			if (_selfdestructin54321)
-				throw new FileBeingDeletedException();
-
-			var internalStreamCount = Interlocked.Increment(ref _internalStreamsCount);
-			if (internalStreamCount > _maxReaderCount)
-				throw new Exception("Unable to acquire reader work item. Max reader count reached.");
-
-			Interlocked.Increment(ref _fileStreamCount);
 			if (_selfdestructin54321) {
 				if (Interlocked.Decrement(ref _fileStreamCount) == 0)
 					CleanUpFileStreamDestruction();
 				throw new FileBeingDeletedException();
 			}
 
-			return CreateInternalReaderWorkItem();
+			Interlocked.Increment(ref _fileStreamCount);
```

Comment:
it looks like this increment operation should be done before `if (_selfdestructin54321) {` just above (as it was in the previous code)

Reply:
Correct. The actual reason for that is `Interlocked.Decrement` which should be placed after increment.